profile-lib/sampler.rkt

```diff
@@ -116,11 +116,25 @@
               (set! next-id (add1 next-id))
               (hash-set! t thd id)
               id)))))
+
+  ;; The sampler thread must hold its watched threads weakly
+  ;; This way if the thread it's watching is dead---which means
+  ;; it'll never run again---then it too can die.
+  (define weak-to-track
+    (let loop ([t to-track])
+      (cond [(thread? t) (make-weak-box t)]
+            [(list? t) (map loop t)]
+            ;; cannot assume that it's a list: we might get other values from
+            ;; a custodian managed list
+            [else t])))
+
   (define (sampler)
     (sleep delay)
+    (define any-left? #f)
```

Comment:
The `any-left?` variable don't get set if `paused` is less than 0, which probably isn't the intent.

Reply:
Overall, it's tempting to suggest that mutable variable is a bad idea here, because it's too easy to overlook what some paths should do.

Reply:
I rewrote to not use a mutable variable; this involved folding a `when` and `unless` into the one big `cond` statement.